File: backend/app/models/results.py
"""
Pydantic results model definitions.
These mirror the TypeScript interfaces in frontend/src/types/results.ts.
"""
from __future__ import annotations

from typing import List, Tuple

from pydantic import BaseModel, field_validator


# ─── Per-Node Displacement ────────────────────────────────────────────────────

class NodeDisplacement(BaseModel):
    node: int
    ux: float   # Horizontal displacement (m)
    uy: float   # Vertical displacement (m)
    rz: float   # Rotation about z-axis (rad)


# ─── Member Internal Forces ───────────────────────────────────────────────────

class MemberForces(BaseModel):
    id: int
    N: List[float]  # Axial force [i-end, j-end] (N)
    V: List[float]  # Shear force [i-end, j-end] (N)
    M: List[float]  # Bending moment [i-end, j-end] (N·m)

    @field_validator('N', 'V', 'M')
    @classmethod
    def validate_two_end_values(cls, v: List[float], info) -> List[float]:
        if len(v) != 2:
            raise ValueError(
                f'{info.field_name} must have exactly 2 values [i-end, j-end], got {len(v)}'
            )
        return v


# ─── Support Reactions ────────────────────────────────────────────────────────

class Reaction(BaseModel):
    node: int
    Fx: float
    Fy: float
    Mz: float


# ─── Plate Stresses ───────────────────────────────────────────────────────────

class PlateStress(BaseModel):
    id: int
    sxx: float        # Normal stress in x (Pa)
    syy: float        # Normal stress in y (Pa)
    sxy: float        # Shear stress (Pa)
    von_mises: float  # Von Mises equivalent stress (Pa)


# ─── Full Analysis Results ────────────────────────────────────────────────────

class AnalysisResults(BaseModel):
    displacements: List[NodeDisplacement]
    member_forces: List[MemberForces]
    reactions: List[Reaction]
    plate_stresses: List[PlateStress]

File: frontend/src/types/results.ts
// ─── Per-Node Displacement ────────────────────────────────────────────────────

export interface NodeDisplacement {
  node: number
  ux: number  // Horizontal displacement
  uy: number  // Vertical displacement
  uz: number  // Out-of-plane displacement
  rx: number  // Rotation about x-axis
  ry: number  // Rotation about y-axis
  rz: number  // Rotation about z-axis
}

// ─── Member Internal Forces ───────────────────────────────────────────────────
// Each component is a tuple [i-end value, j-end value]

export interface MemberForces {
  id: number
  N: [number, number]   // Axial force at [i-end, j-end]
  Vy: [number, number]  // Shear force in y at [i-end, j-end]
  Vz: [number, number]  // Shear force in z at [i-end, j-end]
  T: [number, number]   // Torsion at [i-end, j-end]
  My: [number, number]  // Bending moment about y at [i-end, j-end]
  Mz: [number, number]  // Bending moment about z at [i-end, j-end]
}

// ─── Support Reactions ────────────────────────────────────────────────────────

export interface Reaction {
  node: number
  Fx: number
  Fy: number
  Fz: number
  Mx: number
  My: number
  Mz: number
}

// ─── Plate Stresses ───────────────────────────────────────────────────────────

export interface PlateStress {
  id: number
  sxx: number       // Normal stress in x
  syy: number       // Normal stress in y
  sxy: number       // Shear stress
  von_mises: number // Von Mises equivalent stress
}

// ─── Full Analysis Results ────────────────────────────────────────────────────

export interface AnalysisResults {
  displacements: NodeDisplacement[]
  member_forces: MemberForces[]
  reactions: Reaction[]
  plate_stresses: PlateStress[]
}
